openedx_learning/apps/authoring/units/api.py

```diff
@@ -0,0 +1,208 @@
+"""Units API.
+
+This module provides functions to manage units.
+"""
+
+from django.db.transaction import atomic
+
+from openedx_learning.apps.authoring.containers.models import EntityListRow
+from ..publishing import api as publishing_api
+from ..containers import api as container_api
+from .models import Unit, UnitVersion
+from django.db.models import QuerySet
+
+
+from datetime import datetime
+
+__all__ = [
+    "create_unit",
+    "create_unit_version",
+    "create_next_unit_version",
+    "create_unit_and_version",
+    "get_unit",
+    "get_unit_version",
+    "get_latest_unit_version",
+    "get_user_defined_list_in_unit_version",
+    "get_initial_list_in_unit_version",
+    "get_frozen_list_in_unit_version",
+]
+
+
+def create_unit(
+    learning_package_id: int, key: str, created: datetime, created_by: int | None
+) -> Unit:
+    """Create a new unit.
+
+    Args:
+        learning_package_id: The learning package ID.
+        key: The key.
+        created: The creation date.
+        created_by: The user who created the unit.
+    """
+    with atomic():
+        container = container_api.create_container(
+            learning_package_id, key, created, created_by
+        )
+        unit = Unit.objects.create(
+            container_entity=container,
+            publishable_entity=container.publishable_entity,
+        )
+    return unit
+
+
+def create_unit_version(
+    unit: Unit,
+    version_num: int,
+    title: str,
+    publishable_entities_pk: list[int],
+    created: datetime,
+    created_by: int | None,
+) -> Unit:
+    """Create a new unit version.
+
+    Args:
+        unit_pk: The unit ID.
+        version_num: The version number.
+        title: The title.
+        publishable_entities_pk: The publishable entities.
+        entity: The entity.
+        created: The creation date.
+        created_by: The user who created the unit.
+    """
+    with atomic():
+        container_entity_version = container_api.create_container_version(
+            unit.container_entity.pk,
+            version_num,
+            title,
+            publishable_entities_pk,
+            unit.container_entity.publishable_entity,
+            created,
+            created_by,
+        )
+        unit_version = UnitVersion.objects.create(
+            unit=unit,
+            container_entity_version=container_entity_version,
+            publishable_entity_version=container_entity_version.publishable_entity_version,
+        )
+    return unit_version
+
+
+def create_next_unit_version(
```

Comment:
Regarding this requirement:

_When we pin to different versions of the Container, it indicates a change in how we've defined the Unit, as we explicitly pin a set of versions for the children and may later change to a different set._

This raises a couple of questions. Should we provide methods to pin children, which will then create new versions for the container? Or should we always check whether previous versions were pinned or not, to derive whether versions are now pinned?